traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices_required_capabilities.go

```diff
@@ -329,6 +340,72 @@ func (rc *RequiredCapability) checkServerCap() (error, error, int) {
 	return nil, nil, http.StatusOK
 }
 
+// EnsureTopologyBasedRequiredCapabilities ensures that at least one server per cachegroup
+// in this delivery service's topology has this delivery service's required capabilities.
+func EnsureTopologyBasedRequiredCapabilities(tx *sql.Tx, dsID int, requiredCapabilities []string) (error, error, int) {
+	q := `
+SELECT
+  s.id,
+  c.name,
+  ARRAY_REMOVE(ARRAY_AGG(ssc.server_capability ORDER BY ssc.server_capability), NULL) AS capabilities
+FROM server s
+LEFT JOIN server_server_capability ssc ON ssc.server = s.id
+JOIN cachegroup c ON c.id = s.cachegroup
+JOIN topology_cachegroup tc ON tc.cachegroup = c.name
+JOIN deliveryservice d ON d.topology = tc.topology
+WHERE
+  d.id = $1
+  AND s.cdn_id = (SELECT cdn_id FROM deliveryservice WHERE id = $1)
```

Comment:
Originally, I left this out of the where clause, but it included servers from other CDNs. That was unexpected, since I thought exactly what you thought. I'm not totally sure why, but this condition is necessary.

Reply:
Should there be a test for that? Looks like all of the DSRC tests are on 1 CDN right now.

Reply:
Sure, I can add a test for that.

Reply:
added in https://github.com/apache/trafficcontrol/pull/5093/commits/a108472cfd0db10b0d135459347d62ec484acdf5

Reply:
Tests seem to pass for me with `s.cdn_id = d.cdn_id`

Reply:
Simplified in https://github.com/apache/trafficcontrol/pull/5093/commits/f3a6849cf41a33b36779d0fd6b89b136eb953cf5